analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Branching.LearnConstraint.cs

```diff
@@ -105,21 +121,186 @@ public void Branching_BoolOperation_LearnsBoolConstraint()
         }
 
         [TestMethod]
-        public void Branching_BoolExpression_LearnsBoolConstraint_NotSupported()
+        public void Branching_BoolExpression_LearnsBoolConstraint()
         {
             const string code = @"
 if (boolParameter == true)
 {
-    Tag(""BoolParameter"", boolParameter);
+    Tag(""True"", boolParameter);
+}
+else
+{
+    Tag(""False"", boolParameter);
 }
 bool value;
 if (value = boolParameter)
 {
     Tag(""Value"", value);
 }";
             var validator = SETestContext.CreateCS(code).Validator;
-            validator.ValidateTag("BoolParameter", x => x.Should().BeNull());
-            validator.ValidateTag("Value", x => x.Should().BeNull());
+            validator.ValidateTag("True", x => x.HasConstraint(BoolConstraint.True).Should().BeTrue());
+            validator.ValidateTag("False", x => x.HasConstraint(BoolConstraint.False).Should().BeTrue());
+            validator.ValidateTag("Value", x => x.HasConstraint(BoolConstraint.True).Should().BeTrue());    // Visited only for "true" condition
+        }
+
+        [DataTestMethod]
+        [DataRow("arg == null")]
+        [DataRow("arg == isNull")]
+        [DataRow("null == arg")]
+        [DataRow("isNull == arg")]
+        [DataRow("!!!(arg != null)")]
+        [DataRow("!!!(null != arg)")]
+        [DataRow("!(bool)(object)!!(arg != null)")]
+        [DataRow("!(bool)(object)!!(null != arg)")]
+        public void Branching_LearnsObjectConstraint_CS(string expression)
+        {
+            var code = @$"
+object isNull = null;
+var isObject = new object();
+if ({expression})
+{{
+    Tag(""If"", arg);
+}}
+else
+{{
+    Tag(""Else"", arg);
+}}
+Tag(""End"", arg);";
+            var validator = SETestContext.CreateCS(code, ", object arg").Validator;
```

Comment:
I've extracted this repetitive part into a method in #6040. And also added `ValidateExpectOperation` in it